traitsui/extras/tutor.py

```diff
@@ -873,729 +191,285 @@ def run_code(self):
 
+class TutorialTxtFile(DemoFile):
+    description = Str
 
-class Demo(Lesson):
-    """ Defines a demo, which is a section of a tutorial with both descriptive
-        information and associated Python code which is executed but not
-        shown.
-    """
 
-    #-- Traits View Definitions ----------------------------------------------
-
-    view = View(
-        HSplit(
-            Item('descriptions',
-                 label='Lesson',
-                 style='custom',
-                 show_label=False,
-                 dock='horizontal',
-                 editor=list_editor
-                 ),
-            Item('demo',
-                 id='demo',
-                 style='custom',
-                 show_label=False,
-                 resizable=True,
-                 dock='horizontal',
-                 export='DockWindowShell'
-                 ),
-            id='splitter',
-        ),
-        id='enthought.tutor.demo',
-        handler=LabHandler
+tutorial_text_view = View(
+    Item(
+        "description",
+        label="Description",
+        show_label=False,
+        style="readonly",
+        editor=TextEditor()
     )
-
-#-------------------------------------------------------------------------
-#  'SectionFactory' class:
-#-------------------------------------------------------------------------
+)
 
 
-class SectionFactory(HasPrivateTraits):
-    """ Defines a class that creates Lecture, Lesson or Lab sections (or None),
-        based on the content of a specified directory. None is returned if the
-        directory does not contain any recognized files.
+class TutorialPath(DemoPath):
+    """ A tutorial path also parses file extensions other than py.
     """
 
-    #-- Traits Definitions ---------------------------------------------------
-
-    # The path the section is to be created for:
-    path = Directory
-
-    # The list of files contained in the section:
-    files = List(Str)
-
-    # The parent of the section being created:
-    parent = Instance(ASection)
-
-    # The section created from the path:
-    section = Instance(ASection)
-
-    # The title for the section:
-    title = Str
-
-    # The optional table of contents for the section:
-    toc = List(Str)
-
-    # The list of descriptive items for the section:
-    descriptions = List(ADescriptionItem)
-
-    # The list of code snippet items for the section:
-    snippets = List(CodeItem)
+    def get_children_from_datastructure(self):
+        dirs = []
+        files = []
+        path = self.path
+        for name in os.listdir(path):
+            cur_path = os.path.join(path, name)
+            if os.path.isdir(cur_path):
+                dirs.append(TutorialPath(name=name, parent=self))
+            elif self.use_files:
+                name, ext = os.path.splitext(name)
+
+                # If we have a handler for the file type, invoke it:
+                method = getattr(
```

Comment:
If you wanted to do a bit better of a refactor, replacing "magic method names" with a dictionary mapping file extensions to factory functions would likely make the code much cleaner, and might allow something like having a tutorial be the same base class as a demo but with a slightly different set of registered file handlers, assuming you can't get the codebases completely aligned.

Reply:
The factory functions could be class methods on the appropriate classes being created.